compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -492,255 +536,102 @@ this = mkVar "this"
 self = mkVar "self"
 arg = mkVar "arg"
 res = mkVar "res"
-proxyVar :: TypeVarName
-proxyVar = mkTypeVar "proxy"
-mkProxy :: LF.Type -> LF.Type
-mkProxy = TApp (TVar proxyVar)
+
+convertTemplateDefs :: Env -> ConvertM [Definition]
+convertTemplateDefs env =
+    forM (MS.toList (envTemplateBinds env)) $ \(tname, tbinds) ->
+        DTemplate <$> convertTemplate env tname tbinds
+
+convertTemplate :: Env -> LF.TypeConName -> TemplateBinds -> ConvertM Template
+convertTemplate env tplTypeCon tbinds@TemplateBinds{..}
+    | Just tplTyCon <- tbTyCon
+    , Just fSignatory <- tbSignatory
+    , Just fObserver <- tbObserver
+    , Just fEnsure <- tbEnsure
+    , Just fAgreement <- tbAgreement
+    , tplLocation <- convNameLoc (GHC.tyConName tplTyCon)
+    = withRange tplLocation $ do
+        let tplParam = this
+        tplSignatories <- makeDesugarDFunProjection env fSignatory (`ETmApp` EVar this)
+        tplObservers <- makeDesugarDFunProjection env fObserver (`ETmApp` EVar this)
+        tplPrecondition <- makeDesugarDFunProjection env fEnsure (`ETmApp` EVar this)
+        tplAgreement <- makeDesugarDFunProjection env fAgreement (`ETmApp` EVar this)
+        tplChoices <- convertChoices env tplTypeCon tbinds
+        tplKey <- convertTemplateKey env tplTypeCon tbinds
+        pure Template {..}
+
+    | otherwise =
+        unhandled "Missing required instances in template definition." (show tplTypeCon)
+
+convertTemplateKey :: Env -> LF.TypeConName -> TemplateBinds -> ConvertM (Maybe TemplateKey)
+convertTemplateKey env tname TemplateBinds{..}
+    | Just keyTy <- tbKeyType
+    , Just fKey <- tbKey
+    , Just fMaintainer <- tbMaintainer
+    = do
+        let qtname = Qualified PRSelf (envLFModuleName env) tname
+        tplKeyType <- convertType env keyTy
+        tplKeyBody <- makeDesugarDFunProjection env fKey (`ETmApp` EVar this)
+        tplKeyMaintainers <- makeDesugarDFunProjection env fMaintainer $ \f ->
```

Comment:
This seems to be the only place where you don’t use a section in combination with `makeDesugarDFunProjection` but I don’t really understand why.

Reply:
It's awkward / doesn't really work, because there's two operators in there, `ETyApp` followed by `ETmApp`.